datafusion/core/src/datasource/physical_plan/parquet/row_filter.rs

```diff
@@ -268,90 +259,183 @@ impl<'a> FilterCandidateBuilder<'a> {
-
-                if DataType::is_nested(self.file_schema.field(idx).data_type()) {
-                    self.non_primitive_columns = true;
-                    return Ok(Transformed::new(node, false, TreeNodeRecursion::Jump));
-                }
-            } else if self.table_schema.index_of(column.name()).is_err() {
-                // If the column does not exist in the (un-projected) table schema then
-                // it must be a projected column.
-                self.projected_columns = true;
-                return Ok(Transformed::new(node, false, TreeNodeRecursion::Jump));
+            if let Some(recursion) = self.check_single_column(column.name()) {
+                return Ok(Transformed::new(node, false, recursion));
             }
         }
 
         Ok(Transformed::no(node))
     }
 
-    /// After visiting all children, rewrite column references to nulls if
-    /// they are not in the file schema
     fn f_up(
         &mut self,
         expr: Arc<dyn PhysicalExpr>,
     ) -> Result<Transformed<Arc<dyn PhysicalExpr>>> {
-        // if the expression is a column, is it in the file schema?
         if let Some(column) = expr.as_any().downcast_ref::<Column>() {
             if self.file_schema.field_with_name(column.name()).is_err() {
-                // Replace the column reference with a NULL (using the type from the table schema)
-                // e.g. `column = 'foo'` is rewritten be transformed to `NULL = 'foo'`
-                //
-                // See comments on `FilterCandidateBuilder` for more information
-                return match self.table_schema.field_with_name(column.name()) {
-                    Ok(field) => {
+                // the column expr must be in the table schema
+                return self
+                    .table_schema
+                    .field_with_name(column.name())
+                    .and_then(|field| {
                         // return the null value corresponding to the data type
                         let null_value = ScalarValue::try_from(field.data_type())?;
-                        Ok(Transformed::yes(Arc::new(Literal::new(null_value))))
-                    }
-                    Err(e) => {
-                        // If the column is not in the table schema, should throw the error
-                        arrow_err!(e)
-                    }
-                };
+                        Ok(Transformed::yes(Arc::new(Literal::new(null_value)) as _))
+                    })
+                    // If the column is not in the table schema, should throw the error
+                    .map_err(|e| arrow_datafusion_err!(e));
             }
         }
 
         Ok(Transformed::no(expr))
     }
 }
 
+type ProjectionAndExpr = (Vec<usize>, Arc<dyn PhysicalExpr>);
+
+// Checks if a given expression can be pushed down into `ParquetExec` as opposed to being evaluated
+// post-parquet-scan in a `FilterExec`. If it can be pushed down, this returns None. If it can't be
```

Comment:
this comment and name seems backwards to me (but maybe I am confused) -- the function seems to return `Some(...)` if the expression *could* be pushed down

Reply:
Yes, thanks for pointing that out - I don't know why I had those mixed up for this function. Those should be fixed now.